### PR TITLE
Generate reflection for archetypes

### LINKED ISSUE #6616: Refactor: merge `ArchetypeFieldInfo` into the `reflection` module
https://github.com/rerun-io/rerun/pull/6609 added a reflection API for components (`ComponentName` -> docstrings).

We should use the same interface and system for the `ArchetypeFieldInfo`/`attr.rust.generate_field_info` stuff, which is another form of reflection.

### PR DESCRIPTION
**Pull request by emilk**
### What
* Closes #6616

Replaces the previous code with the new reflection code.

For now I decided that the archetype reflection is still opt-in (with `attr.rust.generate_field_info`) just to reduce the amount of strings bloating our .wasm binary. We can easily change this in the future if we want to.

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using examples from latest `main` build: [rerun.io/viewer](https://rerun.io/viewer/pr/6623?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [rerun.io/viewer](https://rerun.io/viewer/pr/6623?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG
* [x] If applicable, add a new check to the [release checklist](https://github.com/rerun-io/rerun/blob/main/tests/python/release_checklist)!

- [PR Build Summary](https://build.rerun.io/pr/6623)
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

To run all checks from `main`, comment on the PR with `@rerun-bot full-check`.

### CODE REVIEW

File: crates/re_space_view/src/view_property_ui.rs [hunk 1]
```diff
@@ -42,57 +57,38 @@ fn view_property_ui_impl(
...
+            component_results.get_or_empty(field.component_name),
             fallback_provider,
         );
     } else {
         let sub_prop_ui = |ui: &mut egui::Ui| {
-            for component_name in non_indicator_components {
-                let field_info = field_info_per_component.get(component_name);
-                let display_name = field_info
-                    .map_or_else(|| component_name.short_name(), |info| info.display_name);
+            for field in &reflection.fields {
```
[the remark this answers is not in the document]
Comment (Wumpf) [reply crates/re_space_view/src/view_property_ui.rs:81]: nice! I did a similar thing (in spirit) already to the entity visualizer component ui where we use archetype order whereever possible (almost always there's a archetype->visualizer relationship, so the order from there is preserved)

File: crates/re_types_builder/src/codegen/rust/reflection.rs [hunk 2]
```diff
@@ -21,38 +23,48 @@ pub fn generate_reflection(
 
     let mut imports = BTreeSet::new();
     let component_reflection = generate_component_reflection(reporter, objects, &mut imports);
+    let archetype_reflection = generate_archetype_reflection(reporter, objects);
 
     let mut code = format!("// {}\n\n", autogen_warning!());
-    code.push_str("#![allow(unused_imports)]\n");
+    code.push_str("#![allow(clippy::too_many_lines)]\n");
```
Comment (Wumpf) [inline crates/re_types_builder/src/codegen/rust/reflection.rs:29]: huh, news to me that we have this lint active anywhere

File: crates/re_space_view/src/view_property_ui.rs [hunk 3]
```diff
@@ -166,9 +164,10 @@ fn view_property_component_ui(
             .show_hierarchical(ui, singleline_list_item_content)
     };
 
-    if let Some(tooltip) = field_info.map(|info| info.documentation) {
-        list_item_response = list_item_response.on_hover_text(tooltip);
-    }
+    list_item_response = list_item_response.on_hover_ui(|ui| {
+        let id = egui::Id::new((archetype_name, field.display_name));
+        ui.markdown_ui(id, field.docstring_md);
```
Comment (Wumpf) [inline crates/re_space_view/src/view_property_ui.rs:169]: It seems that markdown ui somehow adds a lot of space on top and bottom of the hover ui. Is that intentional? Looks like a bug, orthogonal to this PR

image


Before:
image

**Review by Wumpf — APPROVED** [review]
good to have this in the same place!

File: crates/re_types_builder/src/codegen/rust/reflection.rs (hunk 2 above)
Comment (emilk) [reply crates/re_types_builder/src/codegen/rust/reflection.rs:29]: Good! It means you don't often hit our very high function line limit

https://github.com/rerun-io/rerun/blob/7b359381a987b5efc866cc9b3ef747ec9db003f4/clippy.toml#L24

File: crates/re_space_view/src/view_property_ui.rs (hunk 3 above)
Comment (emilk) [reply crates/re_space_view/src/view_property_ui.rs:169]: It's a bug: https://github.com/lampsitter/egui_commonmark/issues/54

**Pull request merged**